### DISCUSSION TITLE
List initialization.... issue?

### DISCUSSION THREAD

**Discussion by ccarlo88** [dotnet/csharplang#7621] [General] [answered]
Hello,

between all the absurd I have seen in the last years recently VS is asking me to change list initialization from
this:
List<string> tables = new(); 
to this:
List<string> tables = [];

So, my question is: Mixing up a list with an array stuff, is this a bug somehow or is it due to the Python developers not being satisfied enough producing garbage in the Python language that now they want to destroy all other programming languages?

**Comment by sharpchen** [comment, marked as answer]
The `[]` syntax is a new feature from `C# 12`. See https://github.com/dotnet/csharplang/issues/5354?notification_referrer_id=NT_kwDOBJ2IBLMyNjA2MzU1NzU4Ojc3NDMyODM2
And the reason why vs raises the suggestion, it may be related to https://github.com/microsoft/vscode-dotnettools/issues/419. If the project built with no error, that's not a bug because there's a preview version sdk in your machine for sure. If not, that's what confused me about the langVersion and SDKs.

**Comment by huoyaoyuan** [comment]
Collection expression in C# is more like `std::initializer_list` in C++. It can adapt for different collections to provide optimized creation for each collection. It can save people from memorizing `new[] {}`, `new List(capacity)`, `ImmutableArray.Create()` for each collection type.

**Reply by huoyaoyuan** [reply]
Honestly I dislike python very much. However, an *adaptive* collection expression is very useful. The major purpose of this feature is to adapt more collection types with less or zero overhead. It is not a syntax sugar of collection initializer, and will be more powerful.

**Reply by ccarlo88** [reply]
In any (logical) programming language that I know [ ] is used ONLY for arrays. In C, C++. 
I get impressed by the amount of illogical solutions created in C# in the last years, people in a hurry to show something "new" instead of putting some thought on the question: Does it make sense? I am producing any meaningful value on that decision? The fall of C# users should be a metric on this.
I still remember the absurd when it was changed
from:
List<string> abc = new List<string>
to:
List<string> abc = new( )
instead of:
var abc = new List<string>()
As far as I remember, the last one could not be solved due to no one being able to solve a compiler issue where "var" could not be declared out of the method.

**Reply by HaloFour** [reply]
You're free to not use them, and to disable the tooling that suggests using them.  List literal syntax has been a highly requested feature for addition to C# for a long time now, and is designed to work across a large number of collection types, including arrays.

If every feature was stopped by someone loud proclaiming it to be illogical or irrational, C# wouldn't exist.  No programming language would exist.  You're free to disagree with the feature and question it's design, but I'd suggest that the statements you have made will be considered to have violated the code of conduct on this repository.

**Reply by ccarlo88** [reply]
So, the opinion from a couple of people who apparently have nothing to do or maybe their job is keep looking at Github all day long is better than the opinion from people who use C# everyday to work and don't have time to keep following every storm of bad ideas?

**Reply by HaloFour** [reply]
You not liking a feature doesn't make it a bad idea.  This proposal is one of the most upvoted and well supported features in the C# community as a whole.  It had community contribution towards the design and implementation, by people who use C# everyday to work.  The language design took place over several years and the design notes are documented on this repository as to why the syntax was chosen, etc.

**Reply by ccarlo88** [reply]
I am sorry if I am being repetitive but do you believe that the "standard" developer, who passes all day long with tight schedules, comes here on Github, search for every suggestion made and give a feeback?

Maybe there should be a poll (open for months, not 10 days as I have seen before) somewhere else more visible, as an e.g. LinkedIn where the standard worker can say "I don't ever want to see that in my life because now we do have multiple ways to achieve the same thing and [ ] is related to arrays and should not be mixed-up with lists."

**Reply by HaloFour** [reply]
> I am sorry if I am being repetitive but do you believe that the "standard" developer, who passes all day long with tight schedules, comes here on Github, search for every suggestion made and give a feeback?

This is not the only avenue for feedback.

> Maybe there should be a poll (open for months, not 10 days as I have seen before) somewhere else more visible, as an e.g. LinkedIn where the standard worker can say "I don't ever want to see that in my life because now we do have multiple ways to achieve the same thing and [ ] is related to arrays and should not be mixed-up with lists."

The design has been conducted in public for over two years.  It had been proposed in one form or another for another 3-4 years prior to that.  C# governance is not a democracy.  No language is, and for a very good reason.

Again, _you_ not liking something doesn't make it a bad idea.  This is one of the most upvoted and well received proposals to the language, both on this repo and in many other forums.

Short of coming up with a reason that is much better than, "I don't like it," your feedback is very unlikely to gain any traction.  The reasons for using `[]` over `{}`, which is the existing array initializer syntax in C# and C89, is well spelled out in the design notes.  This feature is very explicitly not attempting to be limited to arrays, or lists, or any specific collection type.

**Reply by tannergooding** [reply]
We have many developers heavily involved. This ranges from people who work directly on the .NET team, to MVPs, people who attend the major conferences (.NET Conf, Microsoft BUILD, MVP Summit, etc), to industry professionals, to partners from other major companies using .NET and beyond.

Features often go through years of public design, they get talked about in blogs, shared on social media, presentations given at conferences, and beyond. All with the ability for anyone to also engage here on GitHub or the other venues at any time.

People have been asking for and wanting this feature and it not clicking together for you doesn’t make it bad. [] is also far from just arrays, it is an existing syntax used for collection types in general, it is also used for attributes, fixed sized buffers, and more. It being used this way is not exclusive to C# either and it also happens in other C family languages and in other languages beyond.

The feature exists and it will likely become a heavily used default. You always have the option to setup an .editorconfig that allows you to opt out, but ultimately there are things that only work with the new syntax and avoiding it completely may not always be possible.

— most features are not “just another way to do the same thing”. They bring something unique to the table and address a problem scenario that can’t be solved via integration to an existing feature. Due to how languages work and how things need to be composable with other features, you can often accomplish a few of the same things using different features, but this doesn’t make them equivalent or redundant.

**Reply by ccarlo88** [reply]
> C# governance is not a democracy.

Ok, now you wrote the words that for a long time I have been waiting to read. This is honest and is enough for me to accept. 😄

**Reply by HaloFour** [reply]
> Ok, now you said the words that for a long time I have been waiting to read. This is honest and is enough for me to accept. 😄

That's not to say that feedback isn't considered incredibly important.  Many of the proposals have started or been shaped by community feedback.  Collection/dictionary literals are an example of a feature driven by community feedback, and community members were on the working group that helped to design the feature.

**Reply by jnm2** [reply]
The ability alone to write `[]` instead of `ImmutableArray<(int A, string B, SomeOtherType C)>.Empty` over and over is such a relief.

**Reply by CyrusNajmabadi** [reply]
> Maybe there should be a poll (open for months, not 10 days as I have seen before) somewhere else more visible, 

As Tanner mentioned, we have numerous streams of feedback, and we hear from thousands of developers, teams, groups and companies over the years on the features we work on. This has been one of the most positively received feature we've done in a very long time. 

Visibility does not seem to be a problem. 

Note: as we've mentioned many times in the past, it's a non-goal to please all users. Such a thing turns out to be impossible in practice as the range of customer opinions and views are just too broad (and are often contradictory with each other). As such, we aim to assess the ecosystem as a whole and produce features that in aggregate provide a ton of value to the majority of users.  Unfortunately, fit this feature, you're not in that group. But that's expected that that will happen and we accept that that means a user like you may be unhappy. Just the challenge of doing and sort of work where you have millions of customers.

**Comment by CyrusNajmabadi** [comment]
Brackets are used for far more than arrays in c#. A very important correspondence here which has already shipped is `list patterns`, which uses brackets to already refer to a sequence of patterns to deconstruct a collection into. 

Collection initializers are just the construction form of that.

**Comment by KennethHoff** [comment]
Why limit [] just to arrays? Because it feels nice? What would the alternative for a simple, all-encompassing syntax that works for all collection types be?

**Reply by huoyaoyuan** [reply]
It isn't limited to arrays. The language team has been working on supporting all collection types, including immutable collections which aren't supported by traditional collection initializer.

**Reply by KennethHoff** [reply]
I know. I was asking OP why he felt like [] should be limited to arrays - I can see the misunderstanding though.
